bundles/Billing/Controller/ApiController.php

```diff
@@ -273,9 +280,18 @@ public function exportInvoices(Request $request, BillingService $billingService)
         $writer->setLineEnding("\r\n");
         $writer->setSheetIndex(0);
 
-        $writer->save('php://output');
+        $filesystem = new Filesystem();
+        $filesystem->mkdir($kernel->getProjectDir().'/var/tmp_files/');
+        $tempFilename = $kernel->getProjectDir().'/var/tmp_files/export'.sha1(microtime());
+
+        // Save to temp file.
+        $writer->save($tempFilename);
+
+        $csvOutput = file_get_contents($tempFilename);
+
+        $filesystem->remove($tempFilename);
```

Comment:
This code (with slight variations) is duplicated throughout the project. It may be worthwhile moving it into a service that can take care of the details (with temporary files and stuff) and return the final string.

Maybe `php://memory` (or similar) can be used to skip the temporary file and speed up the generation of the export  (cf. https://www.php.net/manual/en/wrappers.php.php#wrappers.php.memory).

Reply:
I have created a service that hides all the outputting gymnastics.
I will stick with the current temporary file solution, but with the service it is easy to change in the future.